### PR TITLE
Add EEPROM programming support

### PR DESCRIPTION
**Pull request by alexgg**
Depends on https://github.com/balena-os/balena-raspberrypi/pull/1115

### CODE REVIEW

File: layers/meta-balena-raspberrypi/recipes-bsp/rpi-eeprom/rpi-eeprom_git.bb [hunk 1]
```diff
@@ -1,55 +1,122 @@
...
-PV = "20210111"
+SRCREV = "3b393d31ac0f1864420d47028b5703a70ad8da8f"
+PV = "v2023.10.18-2712"
 
 # We use the latest stable version
 # which is available in "stable"
-LATEST_STABLE_PIEEPROM_FW = "2021-01-11"
+LATEST_STABLE_PIEEPROM_FW:raspberrypi4-64 = "2023-05-11"
 VL805_FW_REV = "000138a1"
+LATEST_STABLE_PIEEPROM_FW:raspberrypi5 = "2023-10-18"
```
Comment (alexgg) [inline layers/meta-balena-raspberrypi/recipes-bsp/rpi-eeprom/rpi-eeprom_git.bb:25]: Not tested for RPI5

File: layers/meta-balena-raspberrypi/recipes-bsp/rpi-eeprom/rpi-eeprom_git.bb [hunk 2]
```diff
@@ -63,5 +130,9 @@ INHIBIT_PACKAGE_DEBUG_SPLIT = "1"
 do_deploy[nostamp] = "1"
 
 do_deploy[depends] += " \
-    bootfiles:do_deploy \
+    rpi-bootfiles:do_deploy \
 "
+
+FILES:${PN} = "${libexecdir}/pieeprom-update.sh"
+
+COMPATIBLE_MACHINE = "raspberrypi4-64"
```
Comment (alexgg) [inline layers/meta-balena-raspberrypi/recipes-bsp/rpi-eeprom/rpi-eeprom_git.bb:138]: Might consider raspberrypi5

**Comment by alexgg** [comment]
Manual checks:

[ ] EEPROM updated on RPi4 HUP from older version
[ ] EEPROM not updated on RPI4 HUP from same version
[ ] EEPROM updated on CM4 HUP from older version
[ ] EEPROM not updated on CM4 HUP from same version

File: layers/meta-balena-raspberrypi/recipes-bsp/common/raspberrypi-firmware.inc [hunk 3]
```diff
@@ -0,0 +1,9 @@
+RPIFW_DATE = "20230509~buster"
```
Comment (floion) [inline layers/meta-balena-raspberrypi/recipes-bsp/common/raspberrypi-firmware.inc:1]: I believe we are using that version already: https://github.com/balena-os/balena-raspberrypi/blob/master/layers/meta-balena-raspberrypi/recipes-bsp/bootfiles/rpi-bootfiles.bbappend#L9-L12

Comment (alexgg) [reply layers/meta-balena-raspberrypi/recipes-bsp/common/raspberrypi-firmware.inc:1]: Thanks, this is what happens when code lies in branches for too long. I'll remove this commit from the PR.

File: layers/meta-balena-raspberrypi/recipes-bsp/rpi-eeprom/files/default-config.txt [hunk 4]
```diff
@@ -1,4 +1,7 @@
+[all]
 BOOT_UART=0
 WAKE_ON_GPIO=1
 POWER_OFF_ON_HALT=0
 FREEZE_VERSION=0
+HDMI_DELAY=0
+BOOT_ORDER=0xf25641
```
Comment (acostach) [inline layers/meta-balena-raspberrypi/recipes-bsp/rpi-eeprom/files/default-config.txt:7]: @floion just checking, looking at this [internal thread ](https://balena.zulipchat.com/#narrow/stream/360838-balena-io.2Fos.2Fdevices/topic/rpi.20cm4.20v1.2E1) I'm wondering if 0xf25416 would fit more use-cases?

**Comment by alexgg** [comment]
@resin-jenkins retest this please

File: layers/meta-balena-raspberrypi/recipes-bsp/rpi-eeprom/files/default-config.txt (hunk 4 above)
Comment (alexgg) [reply layers/meta-balena-raspberrypi/recipes-bsp/rpi-eeprom/files/default-config.txt:7]: I agree it's probably better - let's do it in a different PR as it needs some specific testing.

**Comment by alexgg** [comment]
lgtm

**Comment by alexgg** [comment]
@resin-jenkins retest this please

**Comment by alexgg** [comment]
@resin-jenkins retest this please

**Review by acostach — APPROVED** [review]

**Comment by alexgg** [comment]
@resin-jenkins retest this please

**Comment by alexgg** [comment]
@resin-jenkins retest this please

**Comment by alexgg** [comment]
Passed tests:
* test-raspberrypi
* test-raspberrypi3
* test-raspberrypi3-64
* test-raspberrypi4-64 
* test-revpi-core-3
* test-rt-rpi-300 
* test-raspberrypi0-2w-64

**Comment by alexgg** [comment]
@resin-jenkins retest this please

**Comment by alexgg** [comment]
@resin-jenkins retest this please

**Comment by alexgg** [comment]
Devices with required tests that didn't pass:

* test-raspberrypi4-unipi-neuron
* test-revpi-connect

**Pull request merged**